### PR TITLE
Lesson progress, feedback, and result functionalities

### PR DESCRIPTION
**Pull request by inahsanjuan**
- Enable user to start a new lesson
- Lesson can display progress
- Enable user to select choices, and view results
- Finished lesson recorded in user activities

### CODE REVIEW

File: app/Category.php [hunk 1]
```diff
@@ -18,4 +20,40 @@ public function lessons()
...
+
+    public function excerpt()
+    {
+        $str = $this->description;
+        $cont = '';
+
+        if (strlen($str) > config('constants.excerpt_length')) {
+            $str = substr($str, 0, strrpos(substr($str, 0, config('constants.excerpt_length')), ' '));
+            $cont = '...';
+        }
```
Comment (bkintanar) [inline app/Category.php]: @inahsanjuan have you tried using `Str::words($str, 10);` for excerpt?

Comment (inahsanjuan) [inline app/Category.php]: I haven't but I will try using that, didn't know there's an Str facade. Thanks

Comment (bkintanar) [inline app/Category.php]: Ah yes. No problem. You can actually check https://laravel.com/api/5.2/Illuminate/Support.html for more Facades available.

File: app/Http/Controllers/User/LessonController.php [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+            ]);
+
+            foreach ($words as $word) {
+                LessonWord::create([
+                    'word_id' => $word->id,
+                    'lesson_id' => $lesson->id,
+                ]);
+            }
+        }
+        return redirect()->route('lesson.show', $lesson->id);
```
Comment (nguyenthanhtung88) [inline app/Http/Controllers/User/LessonController.php]: Need empty line before `return`

File: app/Http/Controllers/User/LessonController.php [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        }
+        return redirect()->route('lesson.show', $lesson->id);  
+    }
+
+    /**
+     * Display the lesson with words and word answers.
+     */
+    public function show(Lesson $lesson)
+    {
+        $lesson_words = LessonWord::where('lesson_id', $lesson->id)->simplePaginate(1);
```
Comment (nguyenthanhtung88) [inline app/Http/Controllers/User/LessonController.php]: variable should be named in `camcelCase`

File: app/Http/Controllers/User/LessonController.php [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    {
+        //
+    }
+
+    /**
+     * Update user's lesson answers.
+     */
+    public function update(Request $request, Lesson $lesson)
+    {
+        $lesson_word = LessonWord::find($request->lesson_word);
```
Comment (nguyenthanhtung88) [inline app/Http/Controllers/User/LessonController.php]: https://github.com/framgia/fels_168/pull/10/files#r76370863

File: app/Http/Controllers/User/LessonController.php [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+        $lesson_word->setAnswer($request->choice);
+
+        if ($request->next_page == null) {
```
Comment (nguyenthanhtung88) [inline app/Http/Controllers/User/LessonController.php]: Why didn't you use `isset()` or `is_null()` function?

File: app/Http/Controllers/User/LessonController.php [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+            $user = Auth::user();
+
+            $lesson->setResult();
+            $lesson->update();
+
+            $activity->recordActivity($user, 'learned', $lesson);
+             
+            return redirect()->route('lesson.results', $lesson->id);
+        }
+        return redirect($request->next_page);
```
Comment (nguyenthanhtung88) [inline app/Http/Controllers/User/LessonController.php]: https://github.com/framgia/fels_168/pull/10/files#r76370863

File: app/Lesson.php [hunk 7]
```diff
@@ -19,4 +20,20 @@ public function user()
...
+    public function lessonWords()
+    {
+        return $this->hasMany(LessonWord::class);
+    }
+
+    public function setResult()
+    {
+        $results = $this->lessonWords->pluck('is_correct')->all();
+
+        $correct_answers = array_where($results, function($key, $value) {
```
Comment (nguyenthanhtung88) [inline app/Lesson.php]: Variable name should be `camelCase`

File: app/Lesson.php [hunk 8]
```diff
@@ -19,4 +20,20 @@ public function user()
...
+            return $value == 1 ? true : false;
```
Comment (nguyenthanhtung88) [inline app/Lesson.php]: I think `return ($value == 1)` is enough because it return true or false

File: app/LessonWord.php [hunk 9]
```diff
@@ -21,8 +22,36 @@ public function word()
...
     {
         return $this->belongsTo(WordAnswer::class);
     }
+
+    public function choices()
+    {
+        $choices = [];
+        $word_answer = $this->word->wordAnswer;
+        $words = Word::where('category_id', $this->lesson->category_id)
+                    ->get()
```
Comment (nguyenthanhtung88) [inline app/LessonWord.php]: Do we need `->get()` here ???

File: app/LessonWord.php [hunk 10]
```diff
@@ -21,8 +22,36 @@ public function word()
...
-    public function answer()
+    public function wordAnswer()
     {
         return $this->belongsTo(WordAnswer::class);
     }
+
+    public function choices()
+    {
+        $choices = [];
+        $word_answer = $this->word->wordAnswer;
```
Comment (nguyenthanhtung88) [inline app/LessonWord.php]: https://github.com/framgia/fels_168/pull/10/files#r76370863

File: app/LessonWord.php [hunk 11]
```diff
@@ -21,8 +22,36 @@ public function word()
...
+
+        return $choices;
+    }
+
+    public function setAnswer($user_answer)
+    {
+        $answer = $this->word->wordAnswer;
+        
+        $this->word_answer_id = $user_answer;
+        $this->is_correct = $this->word_answer_id == $answer->id ? 1 : 0;
```
Comment (nguyenthanhtung88) [inline app/LessonWord.php]: I think `$this->is_correct = ($this->word_answer_id == $answer->id)` is enough

File: app/User.php [hunk 12]
```diff
@@ -56,6 +56,17 @@ public function lessons()
         return $this->hasMany(Lesson::class);
     }
 
+    public function lessonsCount()
+    {
+        $lessons = $this->lessons()->pluck('result')->all();
+        
+        $lessons_taken = array_where($lessons, function($key, $value) {
+            return is_null($value) ? false : true;
```
Comment (nguyenthanhtung88) [inline app/User.php]: `return is_null($value)` is enough

File: app/LessonWord.php (hunk 9 above)
Comment (inahsanjuan) [inline app/LessonWord.php]: @nguyenthanhtung88 If i remove `->get()`, it will not read `except()` method.

Comment (nguyenthanhtung88) [inline app/LessonWord.php]: @inahsanjuan Why didn't you use `->where('id', '<>', $id)` here.
I saw you `get()`, after that `except()` and call `all()`?

Comment (inahsanjuan) [inline app/LessonWord.php]: @nguyenthanhtung88 Okay, I will try that...I used `all()` to return array and not a collection.

Comment (inahsanjuan) [inline app/LessonWord.php]: @nguyenthanhtung88 I used all() to return array after I used except method. https://laravel.com/docs/5.1/collections#method-all

File: app/Category.php [hunk 13]
```diff
@@ -18,4 +21,18 @@ public function lessons()
...
+    public function excerpt()
+    {
+        $str = $this->description;
+        
+        return Str::words($str, config('constants.excerpt_length'));
+    }
+
+    public function lesson()
+    {
+        return $lesson = Lesson::where('user_id', Auth::user()->id)
```
Comment (nguyenthanhtung88) [inline app/Category.php]: Do you need `$lesson =`?

File: app/LessonWord.php [hunk 14]
```diff
@@ -21,8 +22,36 @@ public function word()
...
+        foreach (array_rand($words, 3) as $index) {
+            array_push($choices, $words[$index]);
+        }
+
+        array_push($choices, $wordAnswer);
+
+        return $choices;
+    }
+
+    public function setAnswer($user_answer)
```
Comment (nguyenthanhtung88) [inline app/LessonWord.php]: `$userAnswer`

File: database/seeds/WordsTableSeeder.php [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+     *
+     * @return void
+     */
+    public function run()
+    {
+        $categories = App\Category::all();
+        $faker = Faker\Factory::create();
+        
+        foreach ($categories as $category) {
+            foreach (range(1,10) as $i) {
```
Comment (nguyenthanhtung88) [inline database/seeds/WordsTableSeeder.php]: One space after `,`

File: resources/assets/js/lesson.js [hunk 16]
```diff
@@ -0,0 +1,19 @@
...
+    var choice_items    = $(choices).find('li');
+
+    var selectRadio = function() {
+        var choice  = $(this);
+        var radio   = choice.children('input[type="radio"]');
+        
+        choice_items.removeClass('emphasize');
+        choice.addClass('emphasize');
+
+        console.log(radio);
```
Comment (nguyenthanhtung88) [inline resources/assets/js/lesson.js]: Remove this log

File: resources/views/category/index.blade.php [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+@section('content')
+    @if ( count($categories))
+        <div class="row">
+            @foreach ($categories as $category)
+                <div class="col-md-6">
+                    @if (count($category->words))
+                        <article class="category">
+                            <div class="block">
+                                <h3 class="title pull-left">{{ $category->title }}</h3>
+                                <span class="word-count pull-right">{{ count($category->words) . ' ' . strToLower(trans('title.words')) }}</span>
```
Comment (nguyenthanhtung88) [inline resources/views/category/index.blade.php]: `strToLower()` ???
http://php.net/manual/en/function.strtolower.php

File: resources/views/lesson/partials/sidebar.blade.php [hunk 18]
```diff
@@ -0,0 +1,5 @@
+<div class="sidebar">
+    <h3>{{ $lesson->category->title }}</h2>
+    <p><b>{{ $lesson->lessonWords->count() }} {{ strToLower(trans('title.words')) }}</b></p>
```
Comment (nguyenthanhtung88) [inline resources/views/lesson/partials/sidebar.blade.php]: http://php.net/manual/en/function.strtolower.php

File: resources/views/lesson/results.blade.php [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+
+@section('content')
+    
+    @include('lesson.partials.sidebar', $lesson)
+    
+    <div class="board">
+        <h3>{{ trans('messages.lesson_results', ['result' => $lesson->result, 'count' => count($lesson->lessonWords)]) }}</h3>
+        <br>
+        <table class="table table-striped">
+            @foreach ($lesson->lessonWords as $lesson_word)
```
Comment (nguyenthanhtung88) [inline resources/views/lesson/results.blade.php]: Variable name should be `camelCase`. Please use `$lessonWord`

File: resources/views/lesson/show.blade.php [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+    {{ trans('title.lessons') }}
+@endsection
+
+@section('content')
+
+    @include('lesson.partials.sidebar', $lesson)
+
+    <div class="board">
+        {{ Form::open(['route' => ['lesson.update', $lesson->id], 'method' => 'PATCH']) }}
+            @foreach ($lessonWords as $lesson_word)
```
Comment (nguyenthanhtung88) [inline resources/views/lesson/show.blade.php]: https://github.com/framgia/fels_168/pull/10/files#r76552107

File: resources/views/lesson/show.blade.php [hunk 21]
```diff
@@ -0,0 +1,50 @@
...
+                <div class="clearfix">
+                    <h1 class="lesson-word border-curved emphasize pull-left">{{ $lesson_word->word->content }}</h3>
+                    <ul class="list-style-none choices-container pull-right">
+                        <fieldset>
+                            @foreach ($lesson_word->choices() as $word_answer)
```
Comment (nguyenthanhtung88) [inline resources/views/lesson/show.blade.php]: https://github.com/framgia/fels_168/pull/10/files#r76552107

File: resources/views/lesson/show.blade.php [hunk 22]
```diff
@@ -0,0 +1,50 @@
...
+                    </ul>
+                </div>
+                
+                {{ Form::hidden('lesson_word', $lesson_word->id) }}
+                {{ Form::hidden('next_page', $lessonWords->nextPageUrl()) }}
+                
+                <div class="form-group text-center">
+                    <p class="lesson-progress">{{ $lessonWords->currentPage() }} / {{ $lesson->lessonWords->count() }}</p>
+                    <button class="btn btn-primary" type="submit">
+                        @if ($lessonWords->nextPageUrl() == null)
```
Comment (nguyenthanhtung88) [inline resources/views/lesson/show.blade.php]: Why did't you use `isset()` or `is_null()` function when compare with null?

**Comment by nguyenthanhtung88** [comment]
OK

**Pull request merged**